### DISCUSSION TITLE
some regressions before 4.0

### DISCUSSION THREAD

**Discussion by spamatica** [muse-sequencer/muse#1206] [Team Posts]
Hi Tim and Kybos, as per usual sorry for being so intermittently available... it's the way it is.

Mostly I want to do the 4.0 release now. :) I don't know any big issues but would like to some more targeted testing that I can hopefully start with soon.
I have an old song I wanted to finish for a competition so I have actually been doing some museing again these last few days and I found some issues.

First off, I mentioned in some ticket that Addictive Drums does not remember it's presets. To my surprise it does remember it in this song now! Not sure if there was some recent change or if the  planets by coincidence are aligned.

- I noticed that the Input and Output route dropdown for audio tracks no longer shows the selected column correctly.

- after exiting the pianoroll (with ESC-key at least) the focus does not go back to the arranger but to one of the zoom widgets, so SPACE no longer start/stop the song but slightly zoom the window until you click in the arranger with the mouse.

- And lastly, the new toolbar for default part color does not have a name in the menu when you right click.

**Comment by kybos** [comment]
I fixed points 2 and 3, but don't get point 1 - I see no issues and I don't remember making any changes there. What's exactly the problem?

As for the final release: Yes, I think we should do it as soon as possible. I am finishing some cosmetic issues (also the gradients from the discussion just below), but won't start anything new, just some more testing. 

If I remember right, there are still some nasty bugs affecting the the latency correction. Perhaps we should mark this functionality "experimental" in the settings, to avoid wrong expectations from users? Just an idea, maybe a bad one, as I don't know much about it.

**Comment by spamatica** [comment]
Alright, cool with the fixes!

Maybe it is something in my configuration that causes problems with the route dropdowns then. I can share a screenshot later in any case and will do some more research.

True with the latency correction. The basic functionality is good and I use it a lot but there are some issues, e.g. not playing the first event and loop playback not working correctly. Marking it as experimental seems a good idea.

**Comment by spamatica** [comment]
Tried on another computer with same kubuntu 20.04 and it happens there too, so something seems to have changed.-
![route dropdown issue](https://user-images.githubusercontent.com/4476983/110856842-6f6deb00-82b8-11eb-8052-2286387d33c0.png)

The column for omni is rendered wierdly.

**Comment by terminator356** [comment]
Looks like your text is drawing on top of the check box.
Not respecting the width of the checkbox, maybe?

I think something changed, I think it used to be a check mark, as seen below.
But I'm not sure if the change is from the style sheet or programmatic, I looked tonight
 but could not find where it does that.

Here's audio tracks for me:
![audio_track_route_popup_1](https://user-images.githubusercontent.com/5422362/110893608-4d7a6580-82c4-11eb-9734-418ffeb8b9d7.png)

Here's midi tracks:
![midi_track_route_popup_1](https://user-images.githubusercontent.com/5422362/110893637-5b2feb00-82c4-11eb-93c7-7345d134399c.png)

**Comment by kybos** [comment]
Yes I see, I know that problem...
It's caused by Qt bugs in the menu styling, unfortunately there are many, and there are essential incompatibilities between Qt releases. It's quite a nightmare...
I will check it.

https://bugreports.qt.io/browse/QTBUG-70491
https://bugreports.qt.io/browse/QTBUG-66380
https://bugreports.qt.io/browse/QTBUG-74655
https://bugreports.qt.io/browse/QTBUG-90242
https://bugreports.qt.io/browse/QTBUG-86754
https://bugreports.qt.io/browse/QTBUG-85832
and many others...

**Comment by kybos** [comment]
@spamatica 
Is the audio routes the only menu where this issues occurs? What about e.g. Settings->Follow Song submenu?
I see that Ubuntu 20.04 has Qt 12.8, which is quite the worst case in the area of menu styling...

**Comment by spamatica** [comment]
Yeah, it's the same in Follow Song.
Right, 5.12.8. Probably they recently bumped the minor version, as far as I remember it used to work fine.

That's a bit unfortunate. 
It's possible to read the dialog anyway, when you know what to look for, but the first time it is a bit confusing. 
Do you know which version this particular problem is fixed?

**Comment by spamatica** [comment]
Could the string with the name be prepadded with spaces?

**Comment by kybos** [comment]
> Could the string with the name be prepadded with spaces?

That would break it for all the other releases.

There are 2 things:
1. In Qt 5.12.6., they changed the left padding calculation incompatibly (before it was calculated from the menu frame, after that from the icon if available). Bad enough, but I managed to cope with that.
2. They, however, forgot the check marks. So in menus that contain only check marks, these are ignored in the calculation. I was able to fix menus with only some check marks, introducing an invisible dummy icon. But with menus that have check marks for all entries, no chance... This is I think fixed in 5.13.1, but this fix introduces other bugs again... An incredible mess.

I see that it can't remain so, as Ubuntu 20.04 is an important release and widely used. I will see what I can do. In the worst case, I will have to remove the menu styling completely, but the default is not good, not enough contrast/visibility for some elements etc., and I put quite a lot of work into it...

**Comment by kybos** [comment]
I did my best to fix all these issues also for Qt. 5.12.8. 
Please check.

**Comment by spamatica** [comment]
![route dropdown issue fixed](https://user-images.githubusercontent.com/4476983/111030621-9be24e00-8403-11eb-8a07-40cf8563975e.png)
Nice, looks good on an initial test anyway! :)

**Comment by terminator356** [comment]
Couple of minor graphic issues that are **not** present without a stylesheet:

The metronome dialog's pattern editor list box is missing the selection highlight bar,
 there should be a selection bar (like any list box):
![metronome_dialog_1](https://user-images.githubusercontent.com/5422362/111036565-c5907a80-83ed-11eb-9261-8d7ba4944cef.png)

When the advanced router first opens, all the Jack related lines are highlighted in white.
As soon as the dialog is updated, say by adding/deleting a track or adding/deleting a 
 connection in QJackCtl for example, the white lines disappear.
The highlight is only supposed to appear when a Jack connection is 'Unavailable',
 say by running dummy driver or a different sound card.
The highlight was originally supposed to be red, but I guess the stylesheet makes it white.
Which makes me think that maybe it's not marking them as 'Unavailable', rather it is simply
 neglecting to fill in the correct colour at first:
![advanced_router_1](https://user-images.githubusercontent.com/5422362/111036569-caedc500-83ed-11eb-9446-606551fda778.png)

**Comment by terminator356** [comment]
Back to the routing popups.
As shown in my previous screenshots above:
Why are the midi track routing popup checkboxes showing check marks
 while audio track routing popups show blue check marks? Shouldn't they be consistent?
I noticed that without a stylesheet, all of them are check marks.

**Comment by kybos** [comment]
Yes, they should be consistent, I noticed that too. There must be some difference in the implementation between the audio and midi popup. The styling is supposed to be applied to all QMenu objects in the same way. I'm looking into it.

The other issues are not new, they are already there in the older themes like Ardour, Dark theme. There is a lot of custom painting involved, that's probably the problem...

**Comment by terminator356** [comment]
My immediate reaction was "ownership problem!".
I think all the items are (obviously) owned by the popup.
But some of the popups may be created and opened without ownership
 and destroyed immediately after by whoever called them. I'll see if I can spot anything...

**Comment by kybos** [comment]
Yes, could be something like that.

I've just found and fixed the metronome problem - the custom painted list item covered everything else so it just needed the transparency attribute.

**Comment by kybos** [comment]
The wrong colors in the route dialog should now be fixed as well.

**Comment by terminator356** [comment]
Unfortunately the advanced router colour fix is not good.

Here's how it's supposed to work:

Suppose you have some USB Jack midi ports listed in the router's lower list box,
 and they are connected to MusE midi devices.

If you now unplug your USB midi device, the Jack midi ports shown in the items
 will turn red, indicating 'Unavailable'.

If you now re-plug your USB midi device, the red colour should **disappear** and the 
 **original item colour should reappear**.
That was the purpose of the forced re-setting of the item colour, which also respected
 the alternating items colours.
With this fix, the items no longer re-set their original colours when the USB device is re-plugged.

Yeah, I looked at the code yesterday and thought "why did I do that ugly thing with palettes there?".
Well, now I am reminded that is the reason why.

Is it possible to cancel this fix, and dig our way out of this problem with something which can
 restore the item colour while still respecting stylesheet colours ?

I hope there is a way.
Thanks.

**Comment by terminator356** [comment]
I found what appears to be the solution.
Not sure if this was available when I wrote the code, or if I just didn't see it:
From QTreeWidgetItem::setBackground() help:
"Setting a default-constructed brush will let the view use the default color from the style."

Tested OK with and without stylesheets, plugging and unplugging USB Jack midi device.
The red items come and go OK now.

**Comment by kybos** [comment]
I see. Yes, that looks like a good solution.

**Comment by kybos** [comment]
Now my audio jack outputs are displayed in red and as "unavailable" in the menu, but they are connected and working normally.

**Comment by terminator356** [comment]
Can you post a picture of what it looks like? Thanks.

**Comment by terminator356** [comment]
Hm, if they are showing as 'Unavailable' in the popup menus, then they must be unavailable,
 since I didn't change anything there.
Are you sure you are not running the Pulse or dummy driver?

**Comment by kybos** [comment]
![image](https://user-images.githubusercontent.com/10009541/111533171-c7986780-8766-11eb-9f2b-a193e05a387d.png)

Jack is running and it also shows as JackAudioDevice in MusE.
Really funny, as the soundcard is connected and working.

**Comment by terminator356** [comment]
Hm, actually the router shows there are no connections from Out 1.
The two Jack 'system:playback_x' really do appear to be missing.
What does QJackCtl tell you? Are those two ports really listed as available there?

**Comment by kybos** [comment]
I guess there is something wrong with my system after all. Will try to restart everything.

**Comment by terminator356** [comment]
(Be sure to tell QJackCtl to show the original names OR alias #2. IOW tell it to show
 the ports as "system:playback_x" not the other two forms.)

**Comment by kybos** [comment]
Thanks and sorry for bothering you, there is really a problem with my system. The outputs disappeared completely now (I use Cadence/Carla from kxstudio for Jack control, the output were always available there, now there is nothing). So definitely not a MusE problem.

**Comment by terminator356** [comment]
Hey no problem. I just happen to be here today :-)
Lemme know if any other trouble, things like that get top priority with me !

**Comment by kybos** [comment]
Perhaps it was the recent system/kernel update, or the soundcard is broken... Something is wrong, system restart did not help either.
Too tired now to continue, going to bed...

**Comment by spamatica** [comment]
Hope you got it working again!

I've been doing some MusEing the last week or so and it has mostly been a very positive experience. Took some time and tested the fixed midi recording modes (overdub/replace, normal/mix/replace) and it looks to be working pretty good!

Currently I've got only one minor thing I wanted to ask about. Again regarding the routing dropdown. Hover cannot be distinguished from a checked item. If you are holding CTRL and marking several items this gets a bit confusing. Maybe make it 50% transparent so the underlying check-state shines through?

**Comment by kybos** [comment]
Yes, my sound is working again. There must have been some evil change in the system update as it stopped to work on both my linux mint systems (one with pulse, the other with jack). With some reconfiguration, I got it to work again on both.

The routing popup: I can't reproduce it on my system, the highlighting at hover is always transparent.  Which popup exactly do you mean, audio or midi? (The implementations are very different.) Could you post a screenshot?

**Comment by spamatica** [comment]
Alright, annoying.

About the popup, I quickly checked the midi popups and they seem fine, but with the audio route it looks like this for me:
![route_dropdown_selected](https://user-images.githubusercontent.com/4476983/112035339-bcb34d80-8b3f-11eb-87e1-eeefe3b79290.png)
I have the mouse on the highlighted entry. It looks selected but isn't actually selected.

**Comment by kybos** [comment]
That's really funny, you mean the checkbox is wrongly displayed as active just because the menu item is highlighted?
I hadn't changed anything in the audio popup anyway because it's a standard menu (using QAction) so there was no need (it's the midi popup which is very special and custom painted).

**Comment by spamatica** [comment]
Yeah... fun :/ This qt version is not the best it seems... 
Tried again the midi popup and that does indeed work just fine.

**Comment by kybos** [comment]
It should be fixed now, please try it out.

**Comment by spamatica** [comment]
Yeah, good work! :)
